### PR TITLE
Make deciphering with AUT-key possible for OpenPGP Card >v3.2 (fixes #1352)

### LINKED ISSUE #1352: OpenPGP Card 3.3: using slot 3 for email encryption
### Problem Description
Older OpenPGP Cards (<version 3) have a limitation which forces us to copy the key of a x.509 key-cert pair to slot 2 as well in order to decrypt messages, instead of just using the key of authentications slot. See third point in Notes [here](https://github.com/OpenSC/OpenSC/wiki/OpenPGP-card#pairs-of-key--certificate-from-p12-file).

This limitation is not the case on OpenPGP Card version 3 anymore. Thus, by not using the encryption slot for the x.509 key-cert pair, one would have two slot for PGP keys, e.g. slot one for 'SC' usage and slot two for 'E' usage. Therefore, one could use PGP keys as well as x.509 encryption at the same card.

I could test this solution already with the [OpenPGP-Csp driver](https://github.com/vletoux/OpenPGP-CSP). So it does work theoretically (on the hardware). But only if OpenSC will work the same way, this would be usable for different applications and systems.

### Proposed Resolution
It would be marvelous if OpenSC would detect usage of OpenPGP Card v3 and tries to use the key slot 3 instead of slot 2. Or both depending on where it is stored...
Maybe I can manage to do this, maybe not. At least a hint where to look in the code would help me a lot!

### PR DESCRIPTION
**Pull request by alex-nitrokey**
This PR makes it possible to use authentication key slot for decryption as well. x.509 certificates are often used for more than one use case, but OpenPGP Cards are constructed with 3 subkeys for different use cases.

Therefore, it was necessary to copy the same key used in AUT-key slot in DEC-key slot as well, for being able to decrypt data (for example for S/MIME mails). See [note 3 in the documentation](https://github.com/OpenSC/OpenSC/wiki/OpenPGP-card#pairs-of-key--certificate-from-p12-file).

Since OpenPGP Card v3.3 it is possible to define PSO:DECIPHER operation for AUT-key as well. The hardware changes are supported in OpenSC with help of this PR.

Tested with Nitrokey Pro 2 (OpenPGP Card v3.3)

##### Checklist

- [ ] Documentation is added or updated
- [x] PKCS#11 module is tested
- [ ] Windows minidriver is tested
- [ ] macOS tokend is tested

### CODE REVIEW

**Comment by frankmorgner** [comment]
Checking for 3.3+ is not enough, you need to evaluate the extended capabilities, see  4.4.3.6 of the specification.

And the spec recommends resetting to the default key-ref after usage.

File: src/libopensc/card-openpgp.c [hunk 1]
```diff
@@ -709,6 +710,11 @@ pgp_get_card_features(sc_card_t *card)
 					if ((priv->sm_algo == SM_ALGO_NONE) && (priv->ext_caps & EXT_CAP_SM))
 						priv->sm_algo = SM_ALGO_UNKNOWN;
 				}
+				if (priv->bcd_version > OPENPGP_CARD_3_3) {
```
[the remark this answers is not in the document]
Comment (alex-nitrokey) [reply src/libopensc/card-openpgp.c]: Yes, you are right, thanks!

File: src/libopensc/card-openpgp.c [hunk 2]
```diff
@@ -1875,6 +1881,63 @@ pgp_set_security_env(sc_card_t *card,
...
 
+/** 
+ * set MANAGE SECURITY ENVIRONMENT as documented in 7.2.18 since OpenPGP Card v3.3
+ *
+ * "This optional command (announced in Extended Capabilities) assigns a specific key to a
+ * command. The DEC-key (Key-Ref 2) can be assigned to the command INTERNAL AUTHENTICATE
+ * and the AUT-Key (Key.Ref 3) can be linked to the command PSO:DECIPHER also."
+ *
+ * key: Key-Ref to change (2 for DEC-Key or 3 for AUT-Key) 
+ * command: Usage to set (2 for PSO:DECIPHER or 3 for INTERNAL AUTHENTICATE)
```
[the remark this answers is not in the document]
Comment (alex-nitrokey) [reply src/libopensc/card-openpgp.c]: Alright, I was thinking about it myself, just wasn't sure yet.

**Comment by alex-nitrokey** [comment]
> Checking for 3.3+ is not enough, you need to evaluate the extended capabilities, see 4.4.3.6 of the specification.

This is done directly in [pgp_set_MSE](https://github.com/OpenSC/OpenSC/pull/1446/files#diff-82e2a2daedd56a1a1feede5735ae4609R1905), isn't it? I could evaluate it previously, in pgp_decipher as well, if you like.

> I think you should initialize the card's callback of set_security_env if the card is compliant with 3.3+. This avoids the extra flag and makes this implementation more consistent with the rest of OpenSC.

Honestly, I am not sure how this would look like, yet. I'll try to understand, but if you have an example, this may help. I just checked the ext_cap like it was done for the other capabilities as well...

**Comment by frankmorgner** [comment]
What I meant was something like this:
```diff
diff --git a/src/libopensc/card-openpgp.c b/src/libopensc/card-openpgp.c
index a0eb3494..609b9f26 100644
--- a/src/libopensc/card-openpgp.c
+++ b/src/libopensc/card-openpgp.c
@@ -1869,6 +1869,13 @@ pgp_set_security_env(sc_card_t *card,
                                "Key reference not compatible with "
                                "requested usage");
                }
+               /* For OpenPGP Card >=v3.3, key slot 3 instead of 2 can be used for
+                * deciphering, but this has to be set via MSE beforehand on every
+                * usage (slot 2 is used by default) see section 7.2.18 of the
+                * specification of OpenPGP Card v3.3 */
+               if (priv->bcd_version >= OPENPGP_CARD_3_3 && env->key_ref[0] == 0x02){
+                       pgp_set_MSE(card, 3, 0xb8);
+               }
                break;
        default:
                LOG_TEST_RET(card->ctx, SC_ERROR_INVALID_ARGUMENTS,
@@ -2041,13 +2048,6 @@ pgp_decipher(sc_card_t *card, const u8 *in, size_t inlen,
        apdu.resp = out;
        apdu.resplen = outlen;
 
-       /* For OpenPGP Card >=v3.3, key slot 3 instead of 2 can be used for deciphering,
-        * but this has to be set via MSE beforehand on every usage (slot 2 is used by default)
-        * see section 7.2.18 of the specification of OpenPGP Card v3.3 */
-       if (priv->bcd_version >= OPENPGP_CARD_3_3 && env->key_ref[0] == 0x02){
-               pgp_set_MSE(card, 3, 0xb8);
-       }
-
        r = sc_transmit_apdu(card, &apdu);
        free(temp);
        LOG_TEST_RET(card->ctx, r, "APDU transmit failed");
```
But please forget this suggestion, as I realized that switching back to the default would still require modifying `pgp_decipher()`. I prefer your solution. And yes, you're also right about checking the extended capabilities.

However, there is still a problem in the length checking. You need to add the following:
```diff
diff --git a/src/libopensc/card-openpgp.c b/src/libopensc/card-openpgp.c
index a0eb3494..c5a76656 100644
--- a/src/libopensc/card-openpgp.c
+++ b/src/libopensc/card-openpgp.c
@@ -710,7 +710,7 @@ pgp_get_card_features(sc_card_t *card)
                                        if ((priv->sm_algo == SM_ALGO_NONE) && (priv->ext_caps & EXT_CAP_SM))
                                                priv->sm_algo = SM_ALGO_UNKNOWN;
                                }
-                               if (priv->bcd_version >= OPENPGP_CARD_3_3) {
+                               if (priv->bcd_version >= OPENPGP_CARD_3_3 && (blob->len >= 11)) {
                                        /* v3.3+: MSE for key numbers 2(DEC) and 3(AUT) supported */
                                        if (blob->data[10])
                                                priv->ext_caps |= EXT_CAP_MSE;
```

If you rebase onto master and perform the [release testing](https://github.com/OpenSC/OpenSC/wiki/Smart-Card-Release-Testing) for OpenPGP, we can add this simple change to 0.19.0.

File: src/libopensc/card-openpgp.c [hunk 3]
```diff
@@ -1875,6 +1881,51 @@ pgp_set_security_env(sc_card_t *card,
...
+ *
+ * "This optional command (announced in Extended Capabilities) assigns a specific key to a
+ * command. The DEC-key (Key-Ref 2) can be assigned to the command INTERNAL AUTHENTICATE
+ * and the AUT-Key (Key.Ref 3) can be linked to the command PSO:DECIPHER also."
+ *
+ * key: Key-Ref to change (2 for DEC-Key or 3 for AUT-Key) 
+ * command: Usage to set (0xb8 for PSO:DECIPHER or 0xa4 for INTERNAL AUTHENTICATE)
+ **/
+static int
+pgp_set_MSE(sc_card_t *card, int key, int command)
```
Comment (frankmorgner) [inline src/libopensc/card-openpgp.c]: please use `u8 p2` instead of `int command`

**Comment by alex-nitrokey** [comment]
> If you rebase onto master and perform the release testing for OpenPGP, we can add this simple change to 0.19.0.

That'd be awesome, I try my best to do it right away, thanks.

**Comment by frankmorgner** [comment]
You could try this patch for the problem with the RNG:
```diff
diff --git a/src/libopensc/card-openpgp.c b/src/libopensc/card-openpgp.c
index 43dc9eaf..9dfb3652 100644
--- a/src/libopensc/card-openpgp.c
+++ b/src/libopensc/card-openpgp.c
@@ -1416,6 +1416,9 @@ pgp_get_challenge(struct sc_card *card, u8 *rnd, size_t len)
        if (priv->max_challenge_size > 0 && len > priv->max_challenge_size) {
                len = priv->max_challenge_size;
        }
+       if (priv->max_challenge_size == 0 && len > 8) {
+               len = 8;
+       }
 
        LOG_FUNC_RETURN(card->ctx, iso_ops->get_challenge(card, rnd, len));
 }
```

Is the problem decrypting with gnuk introduced with this PR? We should look at the debug log...

**Comment by alex-nitrokey** [comment]
> Is the problem decrypting with gnuk introduced with this PR? 

As far as I can see, this exists without this PR already! (tried with and without :wink:)

> We should look at the debug log...

Indeed, [here](https://gist.github.com/alex-nitrokey/f0041f6f033dc20b5ed5bef91dfe3bb1) it is

> You could try this patch for the problem with the RNG

Doesn't change the behaviour.

**Comment by frankmorgner** [comment]
The problem is that the input for deciphering (lc) is too long for the "reader":
```
0x7ff9a1583780 09:35:20.834 [opensc-pkcs11] apdu.c:339:sc_check_apdu: Invalid Case 4 short APDU:
cse=04 cla=00 ins=2a p1=80 p2=86 lc=257 le=256
```
please try setting `max_recv_len` and `max_send_len` in `opensc.conf`.

Your log isn't done with the code of your PR, the log lines don't match. Please re-do the log.

If you additionally submit a log for the RNG problem, I can also look at this in more detail.

**Pull request merged**